File: easy/array/word_pattern/word_pattern.go
package main

import (
	"fmt"
	"strings"
)

func wordPattern(pattern string, str string) bool {
	charToWord, wordToChar := make(map[string]string), make(map[string]string)

	words := strings.Split(str, " ")
	if len(pattern) != len(words) {
		return false
	}

	for i := 0; i < len(pattern); i++ {
		char := string(pattern[i])

		wordFromMap, ok1 := charToWord[char]
		charFromMap, ok2 := wordToChar[words[i]]

		if !ok1 && !ok2 {
			charToWord[char] = words[i]
			wordToChar[words[i]] = char
		} else if !ok1 || !ok2 {
			return false
		} else if wordFromMap != words[i] || charFromMap != char {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println("Solution 1 : ", wordPattern("abba", "dog cat cat dog"))
	fmt.Println("Solution 2 : ", wordPattern("abba", "dog cat cat fish"))
	fmt.Println("Solution 3 : ", wordPattern("aaaa", "dog cat cat dog"))
	fmt.Println("Solution 4 : ", wordPattern("abba", "dog dog dog dog"))
}

File: easy/array/word_pattern/word_pattern.py
class Solution:
    def wordPattern(self, pattern: str, str: str) -> bool:
        char_to_word, word_to_char = {}, {}
        words = str.split(" ")
        if len(pattern) != len(words):
            return False

        for i in range(0, len(words)):
            if not pattern[i] in char_to_word and not words[i] in word_to_char:
                char_to_word[pattern[i]] = words[i]
                word_to_char[words[i]] = pattern[i]
            elif not pattern[i] in char_to_word or not words[i] in word_to_char:
                return False
            elif char_to_word[pattern[i]] != words[i] or word_to_char[words[i]] != pattern[i]:
                return False
        return True


s = Solution()
print("Solution 1 : ", s.wordPattern("abba", "dog cat cat dog"))
print("Solution 2 : ", s.wordPattern("abba", "dog cat cat fish"))
print("Solution 3 : ", s.wordPattern("aaaa", "dog cat cat fish"))
print("Solution 4 : ", s.wordPattern("abba", "dog dog dog fish"))
